libraries/AP_Math/matrixN.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ *  Library for operations on matrices of arbitrary dimensions.
+ *
+ *  Based on the MatrixMath library created by Charlie Matlack on 12/18/10.
+ *  Modified from code by RobH45345 on Arduino Forums, taken from unknown source.
+ */
+
+#pragma GCC optimize("O3")
+ 
+#include "matrixN.h"
+
+void matrix_copy(float* A, int32_t n, int32_t m, float* B){
+    int32_t i, j;
+    
+    for (i = 0; i < m; i++){
+        for (j = 0; j < n; j++){
+            B[n * i + j] = A[n * i + j];
+        }
+    }
+}
+
+    
+// Matrix Multiplication Routine
+// C = A * B
+void matrix_mult(float* A, float* B, int32_t m, int32_t p, int32_t n, float* C){
```

Comment:
In the latest commit I've wrapped these functions in a class, but don't really understand the benefit of this. Why not templatize the functions themselves, without a class wrapper?

As for the arrays, I haven't done the conversion because a) it's a change with a big regression potential; I'd rather check this in as is and do this change later if it's really necessary b) I doubt it makes any perf difference. Indexing for static 2D arrays is just syntactic sugar for the i*dim + j, so it shouldn't be faster.